File: src/main/java/kr/co/bookand/backend/bookstore/domain/BookStore.java
package kr.co.bookand.backend.bookstore.domain;

import com.fasterxml.jackson.annotation.JsonIgnore;
import kr.co.bookand.backend.article.domain.ArticleBookStore;
import kr.co.bookand.backend.common.domain.BaseEntity;
import kr.co.bookand.backend.bookmark.domain.BookmarkBookStore;
import kr.co.bookand.backend.common.domain.Status;
import lombok.*;

import javax.persistence.*;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

import static kr.co.bookand.backend.bookstore.domain.dto.BookStoreDto.*;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor
@Builder
public class BookStore extends BaseEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private String name;
    private String address;
    private String businessHours;
    private String contact;
    private String facility;
    private String sns;

    private String introduction;
    private String mainImage;

    @OneToMany(mappedBy = "bookStore", cascade = CascadeType.ALL)
    private List<BookStoreImage> subImages = new ArrayList<>();

    @Enumerated(EnumType.STRING)
    private Status status;

    private int view;
    private int bookmark;

    @OneToMany(mappedBy = "bookStore", cascade = CascadeType.ALL)
    private List<BookStoreTheme> themeList = new ArrayList<>();

    @OneToMany(mappedBy = "bookStore")
    private List<ArticleBookStore> articleBookStoreList = new ArrayList<>();

    @JsonIgnore
    @OneToMany(mappedBy = "bookStore")
    private List<BookmarkBookStore> bookmarkBookStoreList = new ArrayList<>();

    public void updateBookStoreData(BookStoreRequest bookStoreRequest) {
        this.name = bookStoreRequest.name();
        this.address = bookStoreRequest.address();
        this.businessHours = bookStoreRequest.businessHours();
        this.contact = bookStoreRequest.contact();
        this.facility = bookStoreRequest.facility();
        this.sns = bookStoreRequest.sns();
        this.introduction = bookStoreRequest.introduction();
        this.mainImage = bookStoreRequest.mainImage();
        this.status = Status.valueOf(bookStoreRequest.status());
    }

    public void updateBookStoreStatus(Status status) {
        this.status = status;
    }

    public void updateArticleBookStore(ArticleBookStore articleBookStore) {
        if (articleBookStoreList == null) {
            articleBookStoreList = new ArrayList<>();
        }
        this.articleBookStoreList.add(articleBookStore);
    }

    public void removeArticleBookStore(ArticleBookStore articleBookStore) {
        this.articleBookStoreList.remove(articleBookStore);
    }

    public void updateBookStoreView() {
        this.view++;
    }

    public void updateBookStoreBookmark() {
        this.bookmark++;
    }

    public void updateBookStoreUnBookmark() {
        this.bookmark--;
    }

    public void updateBookStoreImage(BookStoreResponse bookStoreDto) {
        this.mainImage = bookStoreDto.mainImage();
    }

    public void updateBookStoreSubImageDelete(String url) {
        this.subImages.removeIf(subImage -> subImage.getUrl().equals(url));
    }

    public void removeBookmarkBookStore(BookmarkBookStore bookmarkBookStore) {
        this.bookmarkBookStoreList.remove(bookmarkBookStore);
    }

    public void updateBookStoreSubImage(List<BookStoreImage> bookStoreImage) {
        if (subImages == null) {
            subImages = new ArrayList<>();
        }
        this.subImages = bookStoreImage;
    }

    public void updateBookStoreTheme(List<BookStoreTheme> bookStoreTheme) {
        if (themeList == null) {
            themeList = new ArrayList<>();
        }
        this.themeList = bookStoreTheme;
    }
}

File: src/main/kotlin/kr/co/bookand/backend/bookstore/domain/Bookstore.kt
package kr.co.bookand.backend.bookstore.domain

import kr.co.bookand.backend.article.domain.IntroducedBookstore
import kr.co.bookand.backend.bookstore.domain.dto.BookstoreRequest
import kr.co.bookand.backend.common.Status
import kr.co.bookand.backend.common.domain.BaseEntity
import java.time.LocalDateTime
import javax.persistence.*

@Entity
class Bookstore(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "bookstore_id")
    var id: Long = 0,
    var name: String,
    var address: String,
    var businessHours: String,
    var contact: String,
    var facility: String,
    var sns: String,

    var latitude: String,
    var longitude: String,

    var introduction: String,
    var mainImage: String,
    @Enumerated(EnumType.STRING)
    var status: Status,
    var view: Int,
    var bookmark: Int,
    var displayedAt: LocalDateTime?,

    @OneToMany(mappedBy = "bookstore", cascade = [CascadeType.ALL], orphanRemoval = true)
    var themeList: MutableList<BookstoreTheme> = mutableListOf(),

    @OneToMany(mappedBy = "bookstore", cascade = [CascadeType.ALL], orphanRemoval = true)
    var imageList: MutableList<BookstoreImage> = mutableListOf(),

    @OneToMany(mappedBy = "article", cascade = [CascadeType.ALL], orphanRemoval = true)
    var introducedBookstoreList: MutableList<IntroducedBookstore> = mutableListOf(),

    ) : BaseEntity() {
    constructor(bookstoreRequest : BookstoreRequest) : this(
        name = bookstoreRequest.name,
        address = bookstoreRequest.address,
        businessHours = bookstoreRequest.businessHours,
        contact = bookstoreRequest.contact,
        facility = bookstoreRequest.facility,
        sns = bookstoreRequest.sns,
        latitude = bookstoreRequest.latitude,
        longitude = bookstoreRequest.longitude,
        introduction = bookstoreRequest.introduction,
        mainImage = bookstoreRequest.mainImage,
        status = Status.VISIBLE,
        view = 0,
        bookmark = 0,
        displayedAt = null
    )

    fun updateBookstoreStatus(status: Status) {
        this.status = status
    }
    fun updateDisplayedAt(displayedAt: LocalDateTime) {
        this.displayedAt = displayedAt
    }

    fun updateBookstoreTheme(theme: BookstoreTheme) {
        themeList.add(theme)
    }

    fun updateBookstoreThemeList(themeList: MutableList<BookstoreTheme>) {
        this.themeList = themeList
    }

    fun updateBookstoreImage(image: BookstoreImage) {
        imageList.add(image)
    }

    fun updateBookstoreImageList(imageList: MutableList<BookstoreImage>) {
        this.imageList = imageList
    }

    fun updateIntroducedBookstore(introducedBookstore: IntroducedBookstore) {
        introducedBookstoreList.add(introducedBookstore)
    }

    fun updateBookstoreData(bookstoreRequest: BookstoreRequest) {
        name = bookstoreRequest.name
        address = bookstoreRequest.address
        businessHours = bookstoreRequest.businessHours
        contact = bookstoreRequest.contact
        facility = bookstoreRequest.facility
        sns = bookstoreRequest.sns
        introduction = bookstoreRequest.introduction
        mainImage = bookstoreRequest.mainImage
    }

    fun removeIntroducedBookstore(introducedBookstore: IntroducedBookstore) {
        introducedBookstoreList.remove(introducedBookstore)
    }

}